test/kafka-upsert-benchmark/mzworkflows.py

```diff
@@ -0,0 +1,165 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import time
+from typing import List
+
+import pg8000
+
+from materialize import ui
+from materialize.mzcompose import (
+    Kafka,
+    Kgen,
+    Materialized,
+    Testdrive,
+    Workflow,
+    WorkflowArgumentParser,
+    Zookeeper,
+    say,
+)
+from materialize.ui import UIError
+
+mz_persist = Materialized(
+    name="materialized_persist",
+    hostname="materialized",
+    options="--persistent-user-tables --persistent-kafka-upsert-source --disable-persistent-system-tables-test",
+)
+
+mz_default = Materialized()
+
+prerequisites = ["zookeeper", "kafka"]
+
+services = [
+    Zookeeper(),
+    Kafka(),
+    mz_default,
+    mz_persist,
+    # This instance of Testdrive is used to initialize the benchmark, and needs to have a
+    # known seed.
+    Testdrive(no_reset=True, seed=1),
+    Kgen(),
+]
+
+
+def kafka_upsert_benchmark_step(
+    w: Workflow,
+    mz_service: str,
+    restart_mz: bool,
+    records_sent: int,
+    num_records: int,
+    num_keys: int,
+    value_bytes: int,
+    timeout_secs: int,
+):
+    if restart_mz:
+        w.kill_services(services=[mz_service], signal="SIGKILL")
+    w.run_service(
+        service="kgen",
+        command=" ".join(
+            [
+                f"--num-records={num_records}",
+                "--keys=random",
+                "--key-min=0",
+                f"--key-max={num_keys}",
+                "--values=bytes",
+                f"--max-message-size={value_bytes+1}",
+                f"--min-message-size={value_bytes}",
+                "--topic=testdrive-load-test-1",
+                "--quiet",
+            ]
+        ),
+    )
+
+    if restart_mz:
```

Comment:
The first measurement reported is always 1s , which I believe is due to the way events are ordered:
1. kgen
2. start Mz
3. do the SELECT
By the time step 3  is performed, Mz has been up and running and ingesting, which causes that measurement to be incorrect. Here is the log:

```
> CREATE MATERIALIZED VIEW num_records_ingested AS SELECT SUM(max_offset) FROM max_offsets;
rows match; continuing at ts 1641369900.6116853
testdrive completed successfully.
Killing kafka-upsert-benchmark_materialized_1 ... done
Creating kafka-upsert-benchmark_kgen_run ... done
Using 8 threads...
Starting kafka-upsert-benchmark_materialized_1 ... done
C> waiting for dbname=materialize host=localhost port=56467 user=materialize password=postgres to handle 'SELECT 1'success!
C> waiting for materialize to handle 'SELECT * FROM num_records_ingested'C> materialize has not ingested 200000 records, got: 149700
C> query result: 200000 after 1.0029516108334064
success!
```

So Mz has ingested 149700 records already by the time the `SELECT` is issued.